File: Swift/Binary Search/2110.swift
let input = readLine()!.split(separator: " ").map{ Int($0)! }
var (N, C, arr) = ( input[0], input[1], [Int]() )
for _ in 0..<N {
  arr.append(Int(readLine()!)!)
}
arr = arr.sorted()
var (left, right, ans) = (1, arr.last!-arr[0], 0)
while(left <= right) {
  let mid = (left+right)/2
  var (start, cnt) = (0, 1)
  for i in 0..<arr.count {
    if arr[i]-arr[start] >= mid {
      (start, cnt) = (i, cnt+1)
    }
  }
  if cnt >= C {
    if mid > ans {
      ans = mid
    }
    left = mid + 1
  }
  else {
    right = mid - 1
  }
}
print(ans)

File: JavaScript/시험준비/2110.js

const arr = []
function getArray(input, N) {
    for ( let i = 0; i < N; i++ ) {
        arr.push(Number(input[i].trim()))
    }
    arr.sort((a, b) => { return a - b })
}

function calculate(M, C) {
    let ans = 0
    let [s, e] = [0, 9000000000]
    while (s <= e) {
        const mid = Math.floor((s+e)/2)
        let cnt = 0
        let [idx, minP] = [0, 0]
        while(idx < M) {
            if ( minP <= arr[idx] ) {
                minP = arr[idx] + mid
                cnt++
            }
            else {
                idx++
            }
            if (cnt >= C) {
                if ( ans < mid ) { ans = mid }
                break
            }
        }
        if ( cnt >= C ) {
            s = mid + 1
        }
        else {
            e = mid - 1
        }
    }
    console.log(ans)
}

function solution() {
    const fs = require('fs')
    const input = fs.readFileSync('input.txt', 'utf-8').trim().split("\n")

    const [M, C] = input[0].trim().split(" ").map(Number)
    getArray(input.slice(1), M)
    calculate(M, C)
}

solution()